### ISSUE TITLE
Local optimality for alternatives

### ISSUE THREAD

**Issue by genadz** [valhalla/valhalla#3203]
Right now only 2 criteria  are implemented to check if a route can be an alternative: **bounded stretch** and **limited sharing**. There is one more classic criteria that is not implemented: **local optimality** - it helps to filter out routes with useless detours. Here are examples that highlight the problem:

| route with alternatives | suboptimal detour |
| --- | --- |
| ![Screenshot from 2021-07-08 15-24-02](https://user-images.githubusercontent.com/9436623/124923681-5e817c00-e003-11eb-9c55-407710f0192a.png) | ![Screenshot from 2021-07-08 15-19-11](https://user-images.githubusercontent.com/9436623/124923661-57f30480-e003-11eb-8b2e-2abb81e5a27b.png) |
| ![Screenshot from 2021-07-08 15-27-11](https://user-images.githubusercontent.com/9436623/124924027-b02a0680-e003-11eb-8d52-5034e19636f0.png) | ![Screenshot from 2021-07-08 15-30-14](https://user-images.githubusercontent.com/9436623/124924051-b4562400-e003-11eb-8c08-980940a7b65d.png) |

**Comment by genadz** [comment]
I've read some papers about alternatives, so looks like there are 2 possible ways how local optimality can be checked:
1. Direct **"shortest path call"** for each possible candidate.
2. Overlap search trees (forward and reverse) in order to build plateau. Calculate plateau width for each possible candidate to detect local optimality coefficient.

**Comment by genadz** [comment]
### Direct "shortest path call" for each possible candidate

This is the most straightforward approach that says: if we want to check if a subpath is a locally optimal we should make a "shortest path call" with the same origin and destination and check if we get the same path. 

In our case each route is an intersection of forward and reverse searches. So, let's consider such route **P** that consists of 2 parts: **S->V** and **V->T** (**S** - origin, **T** - destination, **V** - intersection). We know, that each part is an optimal route. The only concern is about a subpath around the intersection point **V**. What we do is step back from the intersection point **V** by the distance **T * length(Opt)** in both directions. Suppose it corresponds to the points **X** (on the forward path **S->V**) and **Y** (on the reverse path **V->T**). So, now we build a shortest path from **X** to **Y** and check if it matches with the subpath **X->V->Y** of the route **P**.
![Screenshot from 2021-07-09 11-57-46](https://user-images.githubusercontent.com/9436623/125052740-f8e9ca00-e0ac-11eb-9df3-9f7081adbb64.png)

**Comment by kevinkreiser** [comment]
yeah we decided not to implement this because it was too expensive and we could filter out most of them based on the sharing. in the examples above are you doing 3 alternate? i would have expected sharing to reject the tiny detour in the 3rd alternate

**Comment by genadz** [comment]
> yeah we decided not to implement this because it was too expensive and we could filter out most of them based on the sharing. in the examples above are you doing 3 alternate? i would have expected sharing to reject the tiny detour in the 3rd alternate

In the examples from the description of the issue I search for 2 alternates, in the examples above 1. Also in the images on the right I added locally optimal path that was checked.

**Comment by kevinkreiser** [comment]
Sorry I misworded that I meant 3rd route in total not 3rd alternate
